apps/aefate/src/aefa_data.erl

```diff
@@ -0,0 +1,117 @@
+%% First draft of FATE data representation.
+%% Very likely to change.
+%%
+-include("aefa_data.hrl").
+
+-module(aefa_data).
+
+-type fate_integer() :: ?FATE_INTEGER_T.
+-type fate_boolean() :: ?FATE_BOOLEAN_T.
+-type fate_nil()     :: ?FATE_NIL_T.
+-type fate_list(T)   :: ?FATE_LIST_T(T).
+-type fate_unit()    :: ?FATE_UNIT_T.
+-type fate_map()     :: ?FATE_MAP_T.
+-type fate_string()  :: ?FATE_STRING_T.
+-type fate_address() :: ?FATE_ADDRESS_T.
+-type fate_variant(T):: ?FATE_VARIANT_T(T).
+
+-type fate_void()    :: ?FATE_VOID_T.
+
+-type fate_tuple(T) :: fate_unit()
+                     | ?FATE_TUPLE_T({T})
+                     | ?FATE_TUPLE_T({T,T})
+                     | ?FATE_TUPLE_T({T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T,T,T,T,T,T})
+                     | ?FATE_TUPLE_T({T,T,T,T,T,T,T,T,T,T,T,T,T}).
```

Comment:
This doesn't look very nice, and I'm not sure I see what you are trying to achieve here. Why do we need Erlang types for the VM types in the first place? Won't all the code handle values of arbitrary VM types?

Reply:
I know. It does not look nice at all.
It is so that we can separeate fate types with maps from fate types without maps.

The code will handle arbitrary VM types (at least in most cases). The goal is to limit what a valid Fate type is and to get the limitation that map keys can not have maps in the type description.

The types are also used by Thomas quick check model to generate data.